File: ls/ls.hs
-- $ runhaskell ls.hs
import Control.Monad          (liftM)
import Control.Exception as E (catch)
import Data.List              (intersperse)
import System.Directory       (getDirectoryContents)
import System.Environment     (getArgs)

listFiles :: FilePath -> IO String
listFiles dir = E.catch f handler
        where f = fmap (unlines . removeDotDirectories) $ getDirectoryContents dir
              handler e = return $ show (e :: IOError) ++ "\n"

listFilesWithHeader :: FilePath -> IO String
listFilesWithHeader dir = liftM (("[" ++ dir ++ "]\n") ++) $ listFiles dir

removeDotDirectories :: [FilePath] -> [FilePath]
removeDotDirectories = filter (\x -> not $ x `elem` [".", ".."])

main = do args <- getArgs
          case null args of
              True  -> putStr =<< listFiles "."
              False -> mapM_ putStr =<< (mapM listFilesWithHeader args >>= return . intersperse "\n")

File: ls/ls.java
import java.io.File;

public class ls {
	public static void main(String[] args) {
		if (args.length == 0) {
			listFiles(".");
		} else {
			for (int i = 0; i<args.length; i++) {
				String dir = args[i];
				System.out.println("[" + dir + "]");
				listFiles(dir);
				if (i < (args.length-1))
					System.out.println();
			}
		}
	}

	public static void listFiles(final String dir) {
		try {
			File f = new File(dir);
			for (final File entry : f.listFiles()) {
				System.out.println(entry.getName());
			}
		} catch (Exception e) {
			System.out.println(e.getMessage());
		}
	}
}
